Workflow changes for self-signup for paid software plans

nospame opened the pull request:
## Product Description
Inaccessible until later UI changes are implemented, but adds a redirect on account verification if a user is going through the self-signup for paid plans workflow, as well as the ability for that object to be created when a new domain is registered.

## Technical Summary
[SAAS-16027](https://dimagi.atlassian.net/browse/SAAS-16027), [SAAS-16028](https://dimagi.atlassian.net/browse/SAAS-16028), [SAAS-16029](https://dimagi.atlassian.net/browse/SAAS-16029)

- Adds the new `SelfSignupWorkflow` model as well as its basic methods for use. This will track the workflow of a user who is creating a new account with the intent to subscribe to a paid software plan.
- Allows a `SelfSignupWorkflow` object to be created on domain registration if the `is_self_signup` argument is passed in
- Redirects from the account verification link if a self-signup workflow is currently active for the domain
- Adds automated testing for a good chunk of this
- Left a TODO where this will connect with UI changes in a later PR


## Safety Assurance

### Safety story
None of this code is really active _per se_ until we make the UI changes to pass in `is_self_signup`. I've manually tested the model methods locally and added a good chunk of automated testing for the util and view changes.

### Automated test coverage
`corehq.apps.registration.tests.test_utils.TestRequestNewDomain` and
`corehq.apps.registration.tests.test_views`

### QA Plan
This will get holistic QA with [SAAS-16034](https://dimagi.atlassian.net/browse/SAAS-16034) once UI changes are coded and on staging.


### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [x] The migrations in this code can be safely applied first independently of the code

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change


[SAAS-16027]: https://dimagi.atlassian.net/browse/SAAS-16027?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SAAS-16028]: https://dimagi.atlassian.net/browse/SAAS-16028?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SAAS-16029]: https://dimagi.atlassian.net/browse/SAAS-16029?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SAAS-16034]: https://dimagi.atlassian.net/browse/SAAS-16034?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

nospame committed:
Add SelfSignupWorkflow model

nospame committed:
Allow initiating a self-signup workflow on domain creation

nospame committed:
Test is_self_signup flag

nospame committed:
isort imports

nospame committed:
Redirect to select_plan when self-signup workflow is in progress

nospame committed:
Add tests for confirm_domain view

nospame committed:
update migrations.lock

biyeun commented on corehq/apps/registration/models.py:
```diff
@@ -155,3 +156,32 @@ def clear_data_for_username(cls, username):
+
+
+class SelfSignupWorkflow(models.Model):
+    domain = models.CharField(max_length=255)
```
index on domain

biyeun commented on corehq/apps/registration/models.py:
```diff
@@ -155,3 +156,32 @@ def clear_data_for_username(cls, username):
+    domain = models.CharField(max_length=255)
+    initiating_user = models.CharField(max_length=80)
+    created_on = models.DateTimeField(auto_now_add=True)
+    completed_on = models.DateTimeField(null=True)
```
index on completed on

biyeun commented on corehq/apps/registration/tests/test_utils.py:
```diff
@@ -38,13 +39,16 @@ def setUpClass(cls):
 
     def tearDown(self):
-        for domain in [self.domain_sso_test, self.domain_test]:
+        good_domains = [self.domain_sso_test, self.domain_test]
```
nit: `subscribed_domains`

biyeun commented on corehq/apps/registration/tests/test_utils.py:
```diff
@@ -38,13 +39,16 @@ def setUpClass(cls):
+        SelfSignupWorkflow.objects.filter(domain__in=good_domains).delete()
 
-        for test_domain in ['subscription-failed', 'init-default-roles-failed']:
+        bad_domains = ['subscription-failed', 'init-default-roles-failed']
```
nit: `failed_subscribe_domains`

biyeun commented on corehq/apps/registration/views.py:
```diff
@@ -525,7 +526,9 @@ def confirm_domain(request, guid=''):
-        view_name = "dashboard_default"
+        should_select_plan = bool(request.plan.plan.edition == SoftwarePlanEdition.COMMUNITY
+                                  and SelfSignupWorkflow.get_in_progress_for_domain(req.domain))
+        view_name = 'domain_select_plan' if should_select_plan else 'dashboard_default'
```
can also use `DomainDashboardView.urlname`

nospame replied:
> can also use `DomainDashboardView.urlname`

This one results in a circular import somewhere up the chain, which is maybe why it used the plain string in the first place. Going to see if I can figure that one out.

nospame replied:
I ended up using a helper function with local imports, which I don't totally love. Somewhere up the line from `dashboard.views` there's a decorator `active_domains_required` used, which imports `registration_default` from `registration.views` hence the circular dependency. https://github.com/dimagi/commcare-hq/pull/35177/commits/c46d1df1d6ebf6985555b170f617140c614fd08d

biyeun commented:
reviewed as product team. will ✅ after tests pass and feedback addressed

nospame committed:
PR feedback: model changes
- index on domain and completed_on
- remove created_on from query

nospame committed:
PR feedback: use class urlname for confirm_domain redirect

nospame committed:
PR feedback: test updates
- add structure with create_registration_request
- add SelfSignupWorkflow to IGNORE_MODELS
- nit variable renames

mjriley reviewed (approved):
Thanks for the changes. Minor quibble over the index, but can be merged

nospame commented on corehq/apps/registration/models.py:
```diff
@@ -155,3 +156,36 @@ def clear_data_for_username(cls, username):
+
+    class Meta:
+        indexes = [
+            models.Index(fields=["domain", "completed_on"]),
```
Under normal circumstances, a domain should only ever have one self-signup object, since the workflow will be initiated a the time of domain creation. So I'm not sure if the 'What are all the self signups for X domain' is of particular relevance -- unless indexing here would also help with retrieving the 0 or 1 self signups a domain should have (and it might, I'm not too familiar with indexing generally).

mjriley replied:
I see. This does feel like a nitpick now, but if there should only ever be one self-signup per domain, then indexing the `completed_on` field seems unnecessary, since `domain` would give us the quick lookup for that one object. In addition, I could see wanting to make ad-hoc queries to find the status of a domain's self-signup, without regards to the `completed_on` field. So it's probably fine as is, but might be better to just index on `domain`

nospame replied:
Good point - domain is really the important thing here, so that's probably what makes sense to index on https://github.com/dimagi/commcare-hq/pull/35177/commits/41e0db5b51c5c3582e2d6b522de115b569e62ac3

nospame commented on corehq/apps/registration/views.py:
```diff
@@ -548,6 +557,14 @@ def confirm_domain(request, guid=''):
+def _confirm_domain_redirect(edition, domain):
+    from corehq.apps.dashboard.views import DomainDashboardView
+    from corehq.apps.domain.views import SelectPlanView
+    should_select_plan = bool(edition == SoftwarePlanEdition.COMMUNITY
```
It's functionally the same in this case without `bool`, but the difference is ensuring it's set to either `True` or  `False`, where without `bool` it gets set to either the `SelfSignupWorkflow` instance or `None`.

mjriley replied:
Works for me. Seems like a personal preference

nospame committed:
Only index SelfSignupWorkflow on domain
